// express-ts/lib/collection.js
"use strict";
var Collectable = (function () {
    function Collectable(value, props) {
        if (props === void 0) { props = {}; }
        this.props = {};
        this.value = value;
        this.props = props;
    }
    Collectable.prototype.prop = function (key) {
        return this.props[key];
    };
    Collectable.prototype.check = function (key, value) {
        if (!(value instanceof Array)) {
            value = [value];
        }
        for (var i = 0; i < value.length; i++) {
            if (this.prop(key) === value[i]) {
                return this;
            }
        }
        return null;
    };
    Collectable.prototype.checkAll = function (props) {
        for (var key in props) {
            if (this.check(key, props[key]) === null) {
                return null;
            }
        }
        return this;
    };
    return Collectable;
}());
exports.Collectable = Collectable;
var Collection = (function () {
    function Collection(initial) {
        if (initial === void 0) { initial = []; }
        this.items = initial;
    }
    Collection.prototype.filter = function (props) {
        var filtered = this.items.filter((function (collectable) {
            return collectable.checkAll(props);
        }).bind(this));
        return new Collection(filtered);
    };
    Collection.prototype.map = function (callback) {
        var _this = this;
        return this.items.map((function (item) {
            callback.bind(_this)(item, item.value);
        }).bind(this));
    };
    Collection.prototype.push = function (item, props) {
        if (props === void 0) { props = {}; }
        if (item instanceof Collectable) {
            this.items.push(item);
        }
        else {
            this.items.push(new Collectable(item, props));
        }
    };
    Collection.prototype.pull = function (start) {
        if (start === void 0) { start = 0; }
        return this.items.splice(start, 1)[0].value;
    };
    Collection.prototype.reverse = function () {
        return new Collection(this.items.reverse());
    };
    Collection.prototype.count = function () {
        return this.items.length;
    };
    return Collection;
}());
exports.Collection = Collection;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Collection;
